File: src/main/java/tk/skulk/plugin/core/extensions/updatechecker/listeners/OpJoinListener.java
package tk.skulk.plugin.core.extensions.updatechecker.listeners;

import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.player.PlayerJoinEvent;
import org.jetbrains.annotations.NotNull;
import tk.skulk.plugin.core.extensions.updatechecker.UpdateCheckerExtension;
import tk.skulk.plugin.wrapper.BaseListener;

import static tk.skulk.plugin.helpers.MiniMessageHelper.sendMessage;

public final class OpJoinListener extends BaseListener<UpdateCheckerExtension> {
    public OpJoinListener(final @NotNull UpdateCheckerExtension extension) {
        super(extension);
    }

    @SuppressWarnings("ConstantConditions")
    @EventHandler(priority = EventPriority.LOW)
    public void onPlayerJoin(final @NotNull PlayerJoinEvent event) {
        final var player = event.getPlayer();

        if (!player.isOp()) {
            return;
        }

        final var extension = this.getExtension();
        final var plugin = extension.getPlugin();

        final var currentVersion = plugin.getDescription().getVersion();

        plugin.runAsync(() -> {
            // Blocking HTTP request.
            final var latestVersion = extension.getLatestVersion();

            if (currentVersion.equalsIgnoreCase(latestVersion)) {
                return;
            }

            plugin.runSync(() ->
                    sendMessage(player, "orange", '!', "There is a new version available for the plugin: <0>", latestVersion)
            );
        });
    }
}

File: src/main/kotlin/tk/skulk/plugin/core/extensions/updateChecker/listeners/OpJoinListener.kt
package tk.skulk.plugin.core.extensions.updateChecker.listeners

import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.player.PlayerJoinEvent
import tk.skulk.plugin.core.SNPlugin
import tk.skulk.plugin.core.extensions.updateChecker.UpdateCheckerExtension
import tk.skulk.plugin.framework.Listener
import tk.skulk.plugin.util.runAsync
import tk.skulk.plugin.util.runSync
import tk.skulk.plugin.util.sendMessage

class OpJoinListener(extension: UpdateCheckerExtension) : Listener<SNPlugin, UpdateCheckerExtension>(
    extension
) {
    @EventHandler(priority = EventPriority.LOWEST, ignoreCancelled = true)
    fun onPlayerJoin(event: PlayerJoinEvent) {
        val player = event.player

        if (!player.isOp) {
            return
        }

        runAsync {
            val latestVersion = extension.latestVersion

            @Suppress("UnstableApiUsage") if (latestVersion == extension.plugin.pluginMeta.version) {
                return@runAsync
            }

            runSync {
                player.sendMessage(
                    "gold",
                    '!',
                    "There is a new version of the plugin available: <0>",
                    latestVersion
                )
            }
        }
    }
}
